docs/arguments.rst

```diff
@@ -34,166 +37,86 @@ And what it looks like:
 
     invoke(touch, args=['foo.txt'])
 
-Variadic Arguments
-------------------
 
-The second most common version is variadic arguments where a specific (or
-unlimited) number of arguments is accepted.  This can be controlled with
-the ``nargs`` parameter.  If it is set to ``-1``, then an unlimited number
-of arguments is accepted.
+An argument may be assigned a :ref:`parameter type <parameter-types>`. If no type is provided, the type of the default value is used. If no default value is provided, the type is assumed to be STRING.
 
-The value is then passed as a tuple.  Note that only one argument can be
-set to ``nargs=-1``, as it will eat up all arguments.
+.. admonition:: Note on Required Arguments
 
-Example:
+   It is possible to make an argument required by setting ``required=True``.  It is not recommended since we think command line tools should gracefully degrade into becoming no ops.  We think this because command line tools are often invoked with wildcard inputs and they should not error out if the wildcard is empty.
```

Comment:
If you read the full docs, their are some opinionated pieces that look to be written by the original author and correspond to unix command line tool practices. This is one of them so I modified it slightly and left it in place. I was planning on leaving them in place on my first pass and then we could make a decision as a whole on how opinionated we wanted to be.

Reply:
If you ignore the heading on the note which is the almost the same as the paragraph it was in, then it seems the author was talking about required arguments in general.

Reply:
Makes sense. Could we rephrase "into becoming no ops" here? I think what is meant is that only using arguments makes a cli look like ancient magics, right?

Could we still document `nargs=-1` behaviour with `required=True` however? I agree that only using arguments is bad. Though there's a definite use case for having a variadic argument that requires at least one.

Reply:
I think it means that if you invoke the tool with no arguments then it should do nothing rather than failing with an error. I am going to leave in alone for now and open a meta issue on opinionated takes about things.